### PR TITLE
Fix Zoom Jitter

### PR DESCRIPTION
**Pull request by redweasel**
fixes https://github.com/xournalpp/xournalpp/issues/2703

The fix was to make layout calculate the page height/width with double precision, because otherwise on page N a random error of ~N/2 would have been accumulated.

~~I also refactored some of the space conversion code to be less error prone to double-precision-errors. Also with the old code I noticed, that when zooming with the mouse on the last page, it would randomly jump to prior pages. That is fixed with the refactor. I did test a lot of things, but I am sure I left something untested, so testing is highly encouraged!~~

~~If requested I can split this PR into the refactor/jump fix and the jitter fix as they are mostly separate in code.~~

**Edit:** This PR got split.

### CODE REVIEW

File: src/control/zoom/ZoomControl.cpp
Comment (personalizedrefrigerator) [inline src/control/zoom/ZoomControl.cpp:198]: For consistency, it might be nice to change `unscaledPixels` to `this->unscaledPixels`...

File: src/gui/PageView.h
Comment (personalizedrefrigerator) [inline src/gui/PageView.h:136]: Since these have the same documentation strings as the methods above them, would it be reasonable to use the same docstring for both? E.g.

```c++
    /**
     * Returns the width of this XojPageView as displayed
     * on the display taking into account the current zoom
     */
    int getDisplayWidth() const;
    double getDisplayWidthDouble() const;
    /**
     * Returns the height of this XojPageView as displayed
     * on the display taking into account the current zoom
     */
    int getDisplayHeight() const;
    double getDisplayHeightDouble() const;
```

**Comment by David96** [comment]
Thanks a lot, from a quick testrun, zooming appears to be a lot smoother. I'll continue to use this PR and report back in case I find any issues.
Out of curiosity: why did you replace the getVisibleRect calls by gdk_window_get_root_coords? I didn't have any issues with wrongly centered zooming so I assumed the coordinates were correct.

File: src/gui/inputdevices/TouchInputHandler.cpp
Comment (personalizedrefrigerator) [inline src/gui/inputdevices/TouchInputHandler.cpp:136]: ~~This part might warrant a comment. (I'm not very familiar with GTK, though, so it could just be that :)).~~

Consider moving this code into a method in `MainWindow.h/MainWindow.cpp`. I think this will make it easier to understand what's going on here & limit coupling between the input system and GTK.

File: src/control/zoom/ZoomControl.cpp
Comment (personalizedrefrigerator) [inline src/control/zoom/ZoomControl.cpp:122]: Why don't we need `rect.x` and `rect.y` here?

File: src/control/zoom/ZoomControl.cpp
Comment (personalizedrefrigerator) [inline src/control/zoom/ZoomControl.cpp:23]: Please comment this piece! Some developers (like myself) have little experience with GTK! The documentation for `gdk_window_get_root_coords`. Why aren't we using `event->x` and `event->y`?

**Comment by redweasel** [comment]
I replaced `event->x` with `event->x_root` as well, so instead of calculating the offset as double it calculates the center as integers independent of the scroll value. I don't understand where `event->x` comes from and replacing it in this way seemed to fix the page jump problem when zooming with ctrl+mousewheel. But on the other hand, from googling it seemed like `event->x_root` would be in window space, but it was in screen space for me...

**Review by personalizedrefrigerator — COMMENTED** [review]
From a quick test, this seems to work very well! Thank you!

I have a few comments. These are mostly just me not understanding changes -- additional commenting/moving things into methods with descriptive names could be useful!

File: src/gui/inputdevices/TouchInputHandler.cpp
Comment (personalizedrefrigerator) [reply src/gui/inputdevices/TouchInputHandler.cpp:136]: > I replaced `event->x` with `event->x_root` as well, so instead of calculating the offset as double it calculates the center as integers independent of the scroll value. I don't understand where `event->x` comes from and replacing it in this way seemed to fix the page jump problem when zooming with ctrl+mousewheel. But on the other hand, from googling it seemed like `event->x_root` would be in window space, but it was in screen space for me...

I just saw this! It might be worth documenting this here!

File: src/control/zoom/ZoomControl.cpp
Comment (personalizedrefrigerator) [reply src/control/zoom/ZoomControl.cpp:23]: > I replaced `event->x` with `event->x_root` as well, so instead of calculating the offset as double it calculates the center as integers independent of the scroll value. I don't understand where `event->x` comes from and replacing it in this way seemed to fix the page jump problem when zooming with ctrl+mousewheel. But on the other hand, from googling it seemed like `event->x_root` would be in window space, but it was in screen space for me...

Okay! It might be best to factor things related to this into a single place, so it's easier to change/read. Thank you!

**Comment by personalizedrefrigerator** [comment]
On non-PDFs, I'm having an issue with ctrl+mousewheel zooming... Try opening a blank document, adding a few pages, scrolling down, then ctrl+mousewheel zooming in and out. When I did this, the view moved to the top of one of the pages as soon as I started zooming! It doesn't do this in master.


https://user-images.githubusercontent.com/46334387/106765120-53a66380-65ed-11eb-9688-cff5bcd3c594.mp4

**Comment by David96** [comment]
I looked a bit into this event->x vs event->x_root thing, and the difference seems to be that x_root is relative to the window (as in window managed by a window manager, a GdkWindow might be something different and therefore the confusion with the docs?) and event->x/y is relative to the scroll view, so on large documents event->y becomes very big (which if my understanding of mantissa/exponent is correct could result in less precision for double calculations). Furthermore the values don't seem to have double precision for me anyway, but rather the _root coordinates return always something in the form SomeNumber.0 and the other coordinates return some higher precision value but always the same numbers after the "."

Another weird thing I observed which could hint to a bug in the zooming code (and maybe be related to the bug just reported by @personalizedrefrigerator) is that when scrolling at the end of a document with ctrl+scroll the event->y coordinate jumps to a very small number (basically meaning the document was scrolled up) and then back to the high number which it is supposed to be (without me actually seeing any jumping). Looking at the code, one problem I see is that for mouse wheel scroll, the soom sequence is never ended. Therefore the zoomSequenceStart is never properly set after the first mousewheel zooming. One would need to figure out when the scrolling has stopped, maybe listen to a button release event on ctrl? I think this might fix @personalizedrefrigerator's issue.

This is the quickndirty patch I used to observe the coordinates:
```diff
diff --git a/src/control/zoom/ZoomControl.cpp b/src/control/zoom/ZoomControl.cpp
index 9c8713e4..9c16d823 100644
--- a/src/control/zoom/ZoomControl.cpp
+++ b/src/control/zoom/ZoomControl.cpp
@@ -20,6 +20,7 @@ auto onScrolledwindowMainScrollEvent(GtkWidget* widget, GdkEventScroll* event, Z
                         ZOOM_OUT;
         int rx, ry;
         gdk_window_get_root_coords(gtk_widget_get_window(widget), 0, 0, &rx, &ry);
+        printf("y_root: %f, y: %f\n", event->y_root, event->y);
         zoom->zoomScroll(direction, {event->x_root - rx, event->y_root - ry});
         return true;
     }
```

File: src/control/zoom/ZoomControl.cpp
Comment (personalizedrefrigerator) [reply src/control/zoom/ZoomControl.cpp:23]: I don't know why, but reverting this seems to stop the view from jumping at the beginning of a control+wheel... It does, however, make the view scroll down while doing a ctrl+wheel...

File: src/control/zoom/ZoomControl.cpp
Comment (personalizedrefrigerator) [inline src/control/zoom/ZoomControl.cpp:90]: This makes ctrl+wheel zooming so much better!!!! Thank you!

File: src/control/zoom/ZoomControl.cpp
Comment (redweasel) [reply src/control/zoom/ZoomControl.cpp:122]: I changed `startZoomSequence` to use window coordinates not page relative coordinates. AFAIK `rect.x` would make it page relative. I tested it and it works this way.

File: src/control/zoom/ZoomControl.cpp
Comment (redweasel) [reply src/control/zoom/ZoomControl.cpp:23]: I have no experience with GTK as well. Where (which file) would you put the function for translating screen coordinates to window coordinates?

**Comment by redweasel** [comment]
I think I might have fixed the issue with the page jumps now by always starting the zoom-sequence appropriately (?). It was checking for a change in mouse position previously, but that only introduced bugs and wasn't helping with performance at all (which I think might have been the reason it was there in the first place)

Roadmap for Zooming:
- There is some error of ~3 pixels in `getPaddingLeftOfPage()` I think.
- mouse zoom steps are now different -> reflect that in the settings gui strings and code comments
- Fix/Remove the Gtk touch workaround (it is not working for me at all) (bug not introduced here)
- Fix touch drawing/zoom scroll (does zoom and scroll weirdly, when enabled and three fingers are touched down)
- Make layout system independent of zoom level -> no "`unscaledPixels`" between pages

File: src/control/zoom/ZoomControl.cpp
Comment (personalizedrefrigerator) [reply src/control/zoom/ZoomControl.cpp:125]: Okay! Thank you!

File: src/control/zoom/ZoomControl.cpp
Comment (personalizedrefrigerator) [reply src/control/zoom/ZoomControl.cpp:25]: I think [MainWindow](https://github.com/xournalpp/xournalpp/blob/master/src/gui/MainWindow.h) might be reasonable. I don't see anything like this already in that file, though...
It might also make sense to put this in something in `src/util`, but I don't see a file that provides similar/related functionality.

Comment (redweasel) [reply src/control/zoom/ZoomControl.cpp:25]: Can we leave it as is for this PR and make a separate cleanup PR with no functional changes after this one merged? As I pointed out, this is not the only thing that needs cleanup.

Comment (personalizedrefrigerator) [reply src/control/zoom/ZoomControl.cpp:25]: I think that's fine!

File: src/gui/inputdevices/TouchInputHandler.cpp
Comment (David96) [inline src/gui/inputdevices/TouchInputHandler.cpp:141]: This is wrong. The mainWindow here is of type MainWindow, not GtkWindow. We need the gtk window of the scroll widget, which seems to be a bit harder to get. Otherwise this will return {0,0} instead of the right root_coords which causes a little offset in the zoom center. 

This fixes that but without caring about the touch workaround (in that case, another GtkWidget seems to be used in MainWindow). Also, maybe there is another way of getting the right widget/root coords without adding an additional method to MainWindow.
```diff
diff --git a/src/gui/MainWindow.cpp b/src/gui/MainWindow.cpp
index 3e03ee08..6dc30672 100644
--- a/src/gui/MainWindow.cpp
+++ b/src/gui/MainWindow.cpp
@@ -340,6 +340,8 @@ void MainWindow::initHideMenu() {
 
 auto MainWindow::getLayout() -> Layout* { return gtk_xournal_get_layout(GTK_WIDGET(this->xournal->getWidget())); }
 
+auto MainWindow::getWinXournal() -> GtkWidget* { return this->winXournal; }
+
 auto cancellable_cancel(GCancellable* cancel) -> bool {
     g_cancellable_cancel(cancel);
 
diff --git a/src/gui/MainWindow.h b/src/gui/MainWindow.h
index 875ac3fc..bc8245fc 100644
--- a/src/gui/MainWindow.h
+++ b/src/gui/MainWindow.h
@@ -98,6 +98,7 @@ public:
 
     bool isGestureActive();
 
+    GtkWidget* getWinXournal();
 
     /**
      * Disable kinetic scrolling if there is a touchscreen device that was manually mapped to another enabled input
diff --git a/src/gui/inputdevices/TouchInputHandler.cpp b/src/gui/inputdevices/TouchInputHandler.cpp
index 73a0737a..9573ebbe 100644
--- a/src/gui/inputdevices/TouchInputHandler.cpp
+++ b/src/gui/inputdevices/TouchInputHandler.cpp
@@ -138,10 +138,11 @@ void TouchInputHandler::zoomStart() {
     // use screen pixel coordinates for the zoom center
     // as relative coordinates depend on the changing zoom level
     int rx, ry;
-    gdk_window_get_root_coords(*mainWindow, 0, 0, &rx, &ry);
+    gdk_window_get_root_coords(gtk_widget_get_window(mainWindow->getWinXournal()), 0, 0, &rx, &ry);
     center.x -= rx;
     center.y -= ry;

     // When not using touch drawing, we're using a different scrolling method.
     // This requires different centering.
     /*if (!mainWindow->getGtkTouchscreenScrollingEnabled()) {
 ```

**Comment by redweasel** [comment]
Thanks @David96. I included your patch (2bfeaaf fix window coordinates n TouchInputHandler)

I have also fixed the wrong padding calculation.
I also made the zoom steps comply to the settings, so from 100% Zoom, one step will add 10% if thats whats in your settings. The mouse scroll zoom behavior that @personalizedrefrigerator seemed to really like is unchanged by this.

**Comment by David96** [comment]
Looking at it again, it might be better to do something like this instead of modifying MainWindow. This should also handle the touch workaround case.

But in the end, someone actually associated with this project should comment on that.

```diff
diff --git a/src/control/zoom/ZoomControl.cpp b/src/control/zoom/ZoomControl.cpp
index 42243569..18943302 100644
--- a/src/control/zoom/ZoomControl.cpp
+++ b/src/control/zoom/ZoomControl.cpp
@@ -207,6 +207,7 @@ void ZoomControl::addZoomListener(ZoomListener* l) { this->listener.emplace_back
 void ZoomControl::initZoomHandler(GtkWidget* window, GtkWidget* widget, XournalView* v, Control* c) {
     this->control = c;
     this->view = v;
+    this->widget = widget;
     gtk_widget_add_events(widget, GDK_TOUCHPAD_GESTURE_MASK);
     g_signal_connect(widget, "scroll-event", G_CALLBACK(onScrolledwindowMainScrollEvent), this);
     g_signal_connect(widget, "event", G_CALLBACK(onTouchpadPinchEvent), this);
@@ -226,6 +227,8 @@ void ZoomControl::fireZoomRangeValueChanged() {
     }
 }
 
+auto ZoomControl::getWidget() -> GtkWidget* { return this->widget; }
+
 auto ZoomControl::getZoom() const -> double { return this->zoom; }
 
 auto ZoomControl::getZoomReal() const -> double { return this->zoom / this->zoom100Value; }
diff --git a/src/control/zoom/ZoomControl.h b/src/control/zoom/ZoomControl.h
index a83980b8..084c381a 100644
--- a/src/control/zoom/ZoomControl.h
+++ b/src/control/zoom/ZoomControl.h
@@ -169,6 +169,8 @@ public:
     /// Get visible rect on xournal view, for Zoom Gesture
     Rectangle<double> getVisibleRect();
 
+    GtkWidget* getWidget();
+
     void setZoomStep(double zoomStep);
 
     void setZoomStepScroll(double zoomStep);
@@ -192,6 +194,7 @@ private:
     XournalView* view = nullptr;
     Control* control = nullptr;
     std::vector<ZoomListener*> listener;
+    GtkWidget* widget = nullptr;
 
     /**
      * current Zoom value
diff --git a/src/gui/inputdevices/TouchInputHandler.cpp b/src/gui/inputdevices/TouchInputHandler.cpp
index 73a0737a..fcd49ccc 100644
--- a/src/gui/inputdevices/TouchInputHandler.cpp
+++ b/src/gui/inputdevices/TouchInputHandler.cpp
@@ -134,11 +137,10 @@ void TouchInputHandler::zoomStart() {
 
     auto center = (this->priLastAbs + this->secLastAbs) / 2.0;
 
-    auto* mainWindow = inputContext->getView()->getControl()->getWindow();
     // use screen pixel coordinates for the zoom center
     // as relative coordinates depend on the changing zoom level
     int rx, ry;
-    gdk_window_get_root_coords(*mainWindow, 0, 0, &rx, &ry);
+    gdk_window_get_root_coords(gtk_widget_get_window(zoomControl->getWidget()), 0, 0, &rx, &ry);
     center.x -= rx;
     center.y -= ry;
 ```

File: src/gui/Layout.cpp
Comment (Febbe) [inline src/gui/Layout.cpp:207]:
```suggestion
    auto x = borderX;
    auto y = borderY;
```

File: src/gui/Layout.cpp
Comment (Febbe) [inline src/gui/Layout.cpp:263]:
```suggestion
    auto totalWidth = borderX;
```

File: src/gui/Layout.cpp
Comment (Febbe) [inline src/gui/Layout.cpp:269]:
```suggestion
    auto totalHeight = borderY;
```

File: src/gui/Layout.cpp
Comment (Febbe) [inline src/gui/Layout.cpp:305]: Page 0 should have no `XOURNAL_PADDING_BETWEEN` as offset from the left view anchor.
Where comes the `XOURNAL_PADDING_BETWEEN / 2` from?

Comment (Febbe) [reply src/gui/Layout.cpp:305]: Also the `XOURNAL_PADDING_BETWEEN` is the distance between two pages.

File: src/gui/Layout.cpp
Comment (Febbe) [inline src/gui/Layout.cpp:313]: `columnPadding` should be equal to `XOURNAL_PADDING_BETWEEN`

**Review by Febbe — CHANGES_REQUESTED** [review]
I would aprechiate, if you split up your PR, since I am currently not happy with the change from using local positions to global positions.

Also, I don't have much time to review this in detail. So we have to wait for the next release (1.1.1) instead of 1.1. When you split it, we can at least merge the fix before the release of 1.1.0

File: src/gui/Layout.cpp
Comment (redweasel) [reply src/gui/Layout.cpp:313]: My version is correct AFAICT. See the layout code, to understand how this actually plays out.

File: src/gui/Layout.cpp
Comment (redweasel) [reply src/gui/Layout.cpp:305]: My version is correct AFAICT. See the layout code, to understand how this actually plays out.

**Comment by redweasel** [comment]
I heavily disagree on the usage of `auto` here, but here you go, I changed it. The usage of `auto` was after all the root source of this issue, and the reason why nobody before me found it.

I will split this PR up into the layout part and the zoom input part and squash the commits when I'm finished.

**Comment by Febbe** [comment]
@redweasel as I understand the use of integral values for zoomed ranges was the issue. 

"Using auto everywhere" has the advantage, 

 - That it first is always the correct type
 - That there will never be implicit casts
 - That it defers the need of casts to to place the value is required

Also the use of explicit types caused a huge bug on 32 bit and arm64 systems, because they defined size_t different.
So using `auto` is not the problem it's more like, that we allow implicit casts.

**Comment by redweasel** [comment]
We do not (really) allow lossy implicit casts. There is always a compiler warning if an implicit cast could cause problems. The problem is more like: *we don't listen to the compiler enough.*

*(I found a bunch of dead code in xournalpp regarding file attachment loading, using the compiler warnings, introduced by a clang-tidy operation, but since I wasn't able to produce a bug I didn't mention it yet)*

In this particular case it has to be double and double only! If it would be float, we would have a problem right here again even though it behaves the same, the precision would not suffice (so 0.0 better be double on every possible platform out there or this little change to auto already caused a new issue). `auto` just makes entry harder for new developers. I know this project uses auto everywhere so I comply.

Remember `auto` is controversial to talk about (just google it, there are many people on both ends of the spectrum with good reasons. I'm on the anti part of the spectrum. Hope I am still welcome here xD).

**Comment by Febbe** [comment]
@redweasel Yes I also know both sides :D. I mostly don't care much, but in this case (Layout calculation) we had a huge bug because of named but implementation defined types.

Newcomers are also welcome to ask questions if they do not understand things.

For the split of your PR: I'll try to check all pitfalls, if your code performs better, and I do not find bugs, I'll approve it. So we might merge it before 1.1.0.

Regarding the `XOURNAL_PADDING_BETWEEN` when no one changed the Layout behavior it should be still the distance between two pages. But I'll check the code again on my machine. But for the future it would be way easier, if you explain, why your code is correct, instead of claiming it is correct and linking to the source.

 > Hope I am still welcome here xD
 
Of course, you are :D. Should've marked some `auto` changes as nitpicks.
In particular the `double value = 0;` to `auto value = 0.0` which is only a beauty/consistency nitpick.

**Comment by Febbe** [comment]
If you want to improve the hole Layout / zoom scroll code you could make a plan, we want to rewrite it completely, using some sort of matrix transformation instead of separated zooming and scrolling. Also, it could be good to split the Layout into a 2PageLayout / SinglePageLayout / PresentationLayout.

But I am not sure if we should wait until we change to gtk4. The event system will be changed there fundamentally.

**Comment by redweasel** [comment]
I have now only included the relevant changes to Layout in this PR. I will create another PR for the Zoom Control fixes. (If you try this PR you will immediately see the issues). The issue that this PR really fixes now is the jitter and the wrong zoom center in different configurations.

Sorry for not explaining the code for the padding. I just copied the code directly from the Layout algorithm and removed the unneseccary parts. I don't know why it look like this, it's just what the Layout code did.

@Febbe Yes I was thinking about rewriting the Layout/Zoom code using matrices as well. Xournalpp can have some abstraction around the GTK+ events I think. The Zoom/Layout code should not touch any gtk events from its side. I will make a plan as an issue as soon as I get to it.

**Comment by Febbe** [comment]
Looks good so far, but I think, if we chose to use doubles in the Layout, we shoul do it consequently: 

**nitpick**, because it's constexpr: *https://github.com/xournalpp/xournalpp/blob/d2aaf62208d9c991bdd066d222d2b68b2cfe35fc/src/gui/Layout.cpp#L14-L27

change values to 10.0

**bugprone:** implicit cast, integer division
https://github.com/xournalpp/xournalpp/blob/d2aaf62208d9c991bdd066d222d2b68b2cfe35fc/src/gui/Layout.cpp#L237-L238

Use `2.0` as divisor not `2`

https://github.com/xournalpp/xournalpp/blob/d2aaf62208d9c991bdd066d222d2b68b2cfe35fc/src/gui/Layout.cpp#L161-L162

`pc.minWidth` is still a size_t type, consider to change this also to double or calculate the value with a temporary.
```
    pc.minWidth = floor_cast<size_t>(std::accumulate(begin(pc.widthCols), end(pc.widthCols), double(pc.minWidth)));
    pc.minHeight = floor_cast<size_t>(std::accumulate(begin(pc.heightRows), end(pc.heightRows), double(pc.minHeight)));
```

File: src/gui/Layout.cpp
Comment (Febbe) [inline src/gui/Layout.cpp:162]: Currently, you still use a `size_t` as starting value, this results in casting each value of the vectors to be casted to `size_t` and then added to the starting value.

```suggestion
    pc.minWidth = floor_cast<size_t>(std::accumulate(begin(pc.widthCols), end(pc.widthCols), double(pc.minWidth)));
    pc.minHeight = floor_cast<size_t>(std::accumulate(begin(pc.heightRows), end(pc.heightRows), double(pc.minHeight)));
```

**Comment by Febbe** [comment]
@redweasel thank you for your contribution!

**Pull request merged**
